sirepo/sim_api/jupyterhublogin.py

```diff
@@ -33,15 +34,15 @@
 
 
 class API(sirepo.api.Base):
-    @sirepo.api_perm.allow_visitor
+    @sirepo.api_perm.manual_permission_check
```

Comment:
What I think is going on is that this function is doing too much. It should be `@sirepo.api_perm.require_user` which will throw an exception, and then the code in sirepo.jupyterhub would do the parsing of the response. It's all inside the sirepo repo. Then you wouldn't have to catch anything in this module. All the exceptions would cause a Response to go back to the sirepo.jupyterhub.authenticate. The fall through case (creating the juptyerhub user) would also do the right thing.

Reply:
How will sirepo.jupyterhub build the urls? We have access to self.sirepo_uri but all the methods like uri_for_api and local_route access the flask object directly which we can't do because we are outside the flask context.

Reply:
The schema is available, but not quite. It's what's necessary to map URIs from APIs. For now, these mappings could be hardwired in sirepo.jupyterhub, since we need to get this PR out. 

I don't like adding the new API permission. Another approach is to use self.call_api and have a "hidden" api that api_checkAuthJupyterhub calls and wraps in a proper response. One things we could do to fix the SRException mess is to include a header in the response (X-Sirepo-SRException) that contains a serialized version of the exception so the HTML doesn't need to be parsed (which isn't terrible, and could be better encapsulated).

Anyway, I prefer the first approach and to get to work pulling apart the request dependent (external) from the static stuff (schema and args).

Reply:
890681e6a

Reply:
Thoughts on adding sirepo_uri? Then we can re-use the code without adding another place where we handle creation of uris